oeps/oep-0049-django-app-patterns.rst

```diff
@@ -0,0 +1,211 @@
+      """
+      # _get_programs() returns a queryset
+      q_supported_programs = _get_programs().exclude(
+          uuid__in=UNSUPPORTED_PROGRAM_UUIDS
+      )
+
+      if page_size:
+          # passing a queryset to a paginator allows it to stay performant on large tables
+          program_paginator = Paginator(q_supported_programs, page_size)
+          # get_page returns the first page if page is None
+          supported_programs = program_paginator.get_page(page)
+      else:
+          supported_programs = q_supported_programs
+
+      return [
+          ProgramData(
+              uuid=program.uuid,
+              title=program.title,
+              status=program.status
+          )
+          for program in supported_programs
+      ]
+
+.. _data.py:
+data.py
++++++++
+This file should include the public data structures for the app that can be passed between apps without exposing internal features. These should be used instead of sending Django model objects or querysets to apps that call the functions in ``api.py``. This file should not import anything other than stdlib modules, so that it may be imported by any other app without issue. These data objects should be simple objects with all business logic handled by ``api.py``. They may however perform simple validation, as long as it is self-contained (doesn't reach out to database, network, or any code outside of the class)
+
+For example:
+
+.. code-block:: python
+
+
+  from dataclasses import dataclass
+  from enum import Enum
+
+  import attr
+
+  def ProgramStatus(Enum):
+      ACTIVE = "active"
+      RETIRED = "retired"
+
+  @attr.attrs(frozen=True)
+  class ProgramData:
+      uuid: str = attr.attrib(
+          validator=[attr.validators.instance_of(str)]
+      )
+      title: str = attr.attrib(
+          validator=[attr.validators.instance_of(str)]
+      )
+      status: str = attr.attrib(
+          validator=[
+              attr.validators.instance_of(str),
+              attr.validators.in_(ProgramStatus)
+          ]
+      )
+
+  # Alternatively if you don't need field validation, conversion, or other attrs benefits:
+
+  @dataclass(frozen=True)
+  class ProgramData:
+      uuid: str
+      title: str
+      status: str
+
+.. _rest_api:
+rest_api/
++++++++++
+If an app will have it's own REST API, it should live in a folder called ``rest_api`` to distinguish it from the ``api.py`` file used for intra-app communication.
+
+APIs should be versioned and the serializers and permissions associated with that version should be kept inside that version's folder. This prevents breakages when an API needs to be updated.
+
+An example of a common folder structure for a versioned REST API::
+
+  app_name
+  ├── rest_api
+  │   ├── v1
+  │   │   ├── permissions.py
+  │   │   ├── serializers.py
+  │   │   ├── urls.py
+  │   │   └── views.py
+  │   └── urls.py
+  ├── urls.py
+  └── views.py  # existing legacy non-REST APIs
+
+
+.. _signals:
+signals/
```

Comment:
1. Did you mean to give signals its own directory?
2. It sounds like signal handlers are supposed to go in `signals.py`. If I have a long running task that is triggered by a signal handler, should it go in in the same module, or a separate `tasks.py` file?
3. Where is the right place to define a custom signal that your app will emit? It sounds like we probably don't want it in `signals.py`, since that has celery registration side-effects. Maybe `data.py`?